README.md
# TwitterJSON

TwitterJSON makes it very easy to get up and running with the Twitter REST api on iOS devices. 
You can start getting responses in just a few lines of code. It uses the account from the persons device to 
authorize requests. 

[Video Demo/Tutorial](https://www.youtube.com/watch?v=cXkO4Jt9zF4)

## Example

You can start getting results from Twitter with just one method call. For example, to get the main feed of the user: 

```swift
	TwitterJSON.getHomeFeed { (tweets) -> Void in
            
    }
``` 

All methods are class methods of the `TWitterJSON` class, so no need to initilize any objects. In the example above the
completion handler contains an array of TJTweet objects. Each of which represent a tweet. You can now do what you want
with this array of tweets, for example: 

```swift
    TwitterJSON.getHomeFeed { (tweets) -> Void in
        for tweet in tweets {
            println(tweet.user.name)
            println(tweet.text)
        }
    }
```

Thats it!

## Installation

### CocoaPods

[CocoaPods](http://cocoapods.org) is a dependency manager for Cocoa projects. CocoaPods 0.36 adds   supports for Swift and embedded frameworks. You can install it with the following command:

```bash
    $ gem install cocoapods
```

To integrate TwitterJSON into your Xcode project using CocoaPods, specify it in your `Podfile`:

```ruby
    source 'https://github.com/CocoaPods/Specs.git'
    platform :ios, '8.0'
    use_frameworks!

    pod 'TwitterJSON'
```

Then, run the following command:

```bash
    $ pod install
```

## Setting a ViewController

You should set the view controller of TwitterJSON to whatver view controller is active. This is needed to show alerts to the user. For exmple, if they have disallowed access to Twitter from your app in settings, or they have more than one account setup:

```swift
    TwitterJSON.viewController = self
```

## Number of Results

By default all requests that return an array of either tweets or users will return 20 results. This is defined as a static class variable, meaning that when you set it it will remain at that until it is later changed. To change the number of results:

```swift 
    TwitterJSON.numberOfResults = 5 
```

**Note:** Twitters REST API requests have limits.

## Getting Tweets

The following methods get tweets, and will return an array of TJTweet objects.

```Swift
    TwitterJSON.getHomeFeed { (tweets) -> Void in
        //Returns tweets from the users home feed. 
    }
```

```Swift
    TwitterJSON.searchForTweets("Apple", completion: { (tweets) -> Void in
        //Returns tweets containing the given search query.
    })
```

**Note:** All tweets also contain a `TJUser` object which is 
the infomation about the user who posted the tweet. 

## Getting Users

The following methods get users, and will return an array of TJUser objects.

```swift
    TwitterJSON.getFollowersForUser("KyleGoslan", completion: { (users) -> Void in
        //Returns a list of followers following the specified user.
    })
```

```swift
    TwitterJSON.getFollowingForUser("KyleGoslan", completion: { (users) -> Void in
        //Returns a list of people the specified user is following.
    })
```

## Post Requests

The following methods will post data to Twitter and return a bool value to represent success. 

```swift 
    TwitterJSON.postTweet("Hello World", completion: { (success) -> Void in
        //Will post the given string to twitter.
    })
```

```swift
    TwitterJSON.retweet(123456789, completion: { (success) -> Void in
        //Will retweet the tweet with the given ID.
    })
```

```swift
    TwitterJSON.favoriteTweet(123456789, completion: { (success) -> Void in
        //Will favorite the tweet with the given ID.
    })
```

## Images 
    
All TJUser objects come with the users profile image already downloaded. The download request is performed asynchronously whenever you do any request and is comepleted before the objects are returned, so are ready to use straight away in the completion block. Example:

```swift 
    image.image = tweet.user.profileImage
```

The decision to include profile images by defualt was taken because they are probably the most used images. Doing it this way makes your view update instantly as soon as the callback is recieved while the data consumption is still incredibly small. Typically a request containing 20 tweets is ~20kb of data. 

## Requirements

- iOS 8.0+
- Xcode 6.4 


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

## Unreleased
#### Changed
#### Added

## [1.0.0](https://github.com/KyleGoslan/TwitterJSON/releases/tag/1.0.0)
Released on 2015-16-09.
- Extensive rewrite. Old implementation deprecated, see docs. 
    - Changed by [Kyle Goslan](https://github.com/KyleGoslan).


## [0.2.3](https://github.com/KyleGoslan/TwitterJSON/releases/tag/0.2.3)
Released on 2015-12-09.

#### Added
- Expanded the TJTweet class to incude more properties.
    * retweet count
    * hashtags
    * urls
    * user mentions
    - Added by [Kyle Goslan](https://github.com/KyleGoslan).

## [0.2.2](https://github.com/KyleGoslan/TwitterJSON/releases/tag/0.2.2)
Released on 2015-08-09.

#### Added
- All TJUser objects now have the users profile image autmatically downloaded and ready to use. 
     - Added by [Kyle Goslan](https://github.com/KyleGoslan).
- You can now print your twitter apps rate limit to check your useage. 
     - Added by [Kyle Goslan](https://github.com/KyleGoslan).
- Number of properties to TJUser objects.
     - Added by [Kyle Goslan](https://github.com/KyleGoslan).


## [0.2.1](https://github.com/KyleGoslan/TwitterJSON/releases/tag/0.2.1)
Released on 2015-08-07.

#### Changed
- Fixed bug in searchForTweets method that resulted in no errors being returned when there should be.
    - Changed by [Kyle Goslan](https://github.com/KyleGoslan).

## [0.2.0](https://github.com/KyleGoslan/TwitterJSON/releases/tag/0.2.0)
Released on 2015-08-07.

#### Added
- Added error handling for all API method calls. See documentation. 
     - Added by [Kyle Goslan](https://github.com/KyleGoslan).

## [0.1.5](https://github.com/KyleGoslan/TwitterJSON/releases/tag/0.1.5)
Released on 2015-08-05.

#### Added
- Added TJUsers `getFollowingForUser(screenName: String, completion: (users: [TJUser]) -> Void) {..}` method.
    - Added by [Kyle Goslan](https://github.com/KyleGoslan).
- Added TJUsers `getFollowingForUser(screenName: String, completion: (users: [TJUser]) -> Void) {..}` method.
    - Added by [Kyle Goslan](https://github.com/KyleGoslan).
- Added TJTweets `func getTweetsForList(listSlug: String, fromUser user: String, completion: (tweets: [TJTweet]) -> Void) {...}` method.
    - Added by [Kyle Goslan](https://github.com/KyleGoslan).
- Added a number of tests.
    - Added by [Kyle Goslan](https://github.com/KyleGoslan).

## [0.1.4](https://github.com/KyleGoslan/TwitterJSON/releases/tag/0.1.4)
Released on 2015-08-05.

#### Changed
- Removed the TwitterJSON protocol methods so this class no longer has a delegate. Date is now returned in a completion handler.
    - Changed by [Kyle Goslan](https://github.com/KyleGoslan).
- TJTweets & TJUsers objects no longer have delegate methods. Uses completion handlers instead, which makes the library even easier to implement and use.
    - Changed by [Kyle Goslan](https://github.com/KyleGoslan).
    
#### Added
- Completion handlers rather than delegate methods.
    - Added by [Kyle Goslan](https://github.com/KyleGoslan).
- Added tests.
    - Added by [Kyle Goslan](https://github.com/KyleGoslan).

## [0.1.3](https://github.com/KyleGoslan/TwitterJSON/releases/tag/0.1.3)
Released on 2015-08-04.

#### Added
- Added TJUsers class for retrieving list of users.
  - Added by [Kyle Goslan](https://github.com/KyleGoslan).


## [0.1.2](https://github.com/KyleGoslan/TwitterJSON/releases/tag/0.1.2)
Released on 2015-08-04.

#### Added
- Improved documentation in README.
  - Added by [Kyle Goslan](https://github.com/KyleGoslan).
- Added inline documentation for help on method calls inside Xcode.
  - Added by [Kyle Goslan](https://github.com/KyleGoslan).
  
## [0.1.1](https://github.com/KyleGoslan/TwitterJSON/releases/tag/0.1.1)
Released on 2015-08-04.

#### Added
- Core functionality.
  - Added by [Kyle Goslan](https://github.com/KyleGoslan).

## [0.0.1](https://github.com/KyleGoslan/TwitterJSON/releases/tag/0.0.1)
Released on 2015-08-03.

#### Added
- Initial release of TwitterJSON.
  - Added by [Kyle Goslan](https://github.com/KyleGoslan).

TwitterJSON/TJTweet.swift
//
//  TJTweet.swift
//  TwitterJSON
//
//  Created by Kyle Goslan on 04/08/2015.
//  Copyright (c) 2015 Kyle Goslan. All rights reserved.
//

import Foundation
import SwiftyJSON

/**
A single tweet object.

This also contains a TJUser which contains information about the user who posted the tweet.
*/
public class TJTweet {
    
    /**
    The user who the tweet belongs to
    */
    public var user: TJUser!
    
    /**
    The main 140 characters text of the tweet
    */
    public var text: String!
    
    /**
    The ID of the tweet
    */
    public var id: Int!
    
    /**
    How many times the tweet has been favorited
    */
    public var favoriteCount: Int!
    
    /**
    How many times the tweet has been retweeted
    */
    public var retweetCount: Int!
    
    /**
    Array of hashtags in the tweet
    */
    public var hashtags: [String] = []
    
    /**
    Array of urls in the tweet
    */
    public var urls: [String] = []
    
    /**
    Array of users mentioned in the tweet
    */
    public var userMentions: [String] = []
    
    /**
    Initilize with a json object.
    
    The init method parses the json and sets the objets properties.
    */
    init(tweetInfo: JSON) {
        
        if let id = tweetInfo["id"].int {
            self.id = id
        }
        
        if let user = tweetInfo["user"].dictionary {
            self.user = TJUser(userInfo: tweetInfo["user"])
        }
        
        if let text = tweetInfo["text"].string {
            self.text = text
        }
        
        if let favoriteCount = tweetInfo["favorite_count"].int {
            self.favoriteCount = favoriteCount
        }
        
        if let retweetCount = tweetInfo["retweet_count"].int {
            self.retweetCount = retweetCount
        }
        
        if let hashtags = tweetInfo["entities","hashtags"].array {
            for tag in hashtags {
                self.hashtags.append(tag["text"].string!)
            }
        }
        
        if let urls = tweetInfo["entities","urls"].array {
            for url in urls {
                self.urls.append(url["expanded_url"].string!)
            }
        }
        
        if let userMentions = tweetInfo["entities","user_mentions"].array {
            for mention in userMentions {
                self.userMentions.append(mention["screen_name"].string!)
            }
        }

    }
    
}







